File: Programmers/KAKAO/주차 요금 계산.cpp
// 2022 KAKAO BLIND RECRUITMENT level2
#include <string>
#include <vector>
#include <sstream>
#include <map>

using namespace std;

int get_tot_fee(int tot, vector<int> fees) {
    // 기본 요금 안이면
    if (tot <= fees[0]) {
        return fees[1];
    }
    
    else {
        // 소수점이 있는 경우
        if ((tot - fees[0]) % fees[2]) {
            return fees[1] + ((int)(tot - fees[0]) / fees[2] + 1) * fees[3];
        }   
        else {
            return fees[1] + ((int)(tot - fees[0]) / fees[2]) * fees[3];
        }
    }
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    map<string, vector<string>> hash; // 자동정렬을 위해
    
    for (auto record : records) {
        stringstream sstr(record);
        string time, number, state;
        
        sstr >> time >> number >> state;
        hash[number].push_back(time);
    }
    
    for (auto h : hash) {
        // 입출입이 홀수 이면
        if (h.second.size() % 2) {
            h.second.push_back("23:59");
        }
        
        int tot = 0, in_h, in_m, out_h, out_m;
        for (int i = 0; i < h.second.size(); ++i) {
            string car = h.second[i];
            // 0 ,2 ,4 in, 1, 3, 5 out
            if (i % 2) {
                out_h = stoi(car.substr(0, 2));
                out_m = stoi(car.substr(3, 2));
                
                if (out_m < in_m) {
                    tot += ((out_h - in_h - 1) * 60 + (60 + out_m - in_m));
                }
                
                else {
                    tot += ((out_h - in_h) * 60 + (out_m - in_m));
                }
            }
            
            else {
                in_h = stoi(car.substr(0, 2));
                in_m = stoi(car.substr(3, 2));
            }
        }
        
        answer.push_back(get_tot_fee(tot, fees));
    }
    
    return answer;
}

File: Programmers/KAKAO/주차 요금 계산.py
#// 2022 KAKAO BLIND RECRUITMENT level2
import math

def solution(fees, records):
    answer = []
    in_records_cars = dict() # 들어오는 자동차 
    tot_minute = dict()  # 자동차의 총 주차시간
    
    for record in records:
        time, car, state = record.split() # 공백으로 분리
        h, m = time.split(":")
        minute = int(h) * 60 + int(m) # 분으로 환산
        
        if state == "IN":
            in_records_cars[car] = minute;
            
        else:
            # tot_minute 딕셔너리가 존재할 경우
            try:
                tot_minute[car] += minute - in_records_cars[car]
            # 추가 되는 경우
            except:
                tot_minute[car] = minute - in_records_cars[car]
            del in_records_cars[car]
    
    # 0000에 나갔을 경우가 존재
    for car, minute in in_records_cars.items():
        # tot_minute 딕셔너리가 존재할 경우
        try:
            tot_minute[car] += 23 * 60 + 59 - minute
        # 추가 되는 경우
        except:
            tot_minute[car] = 23 * 60 + 59 - minute
    
    # 정렬하면서 총 요금 계산
    for car, minute in sorted(tot_minute.items()):
        # 나머지가 존재하면 +1
        # 음수인 경우는 그냥 기본 요금
        fee = fees[1] + math.ceil(max(0, (minute - fees[0])) / fees[2]) * fees[3]
        answer.append(fee)
        
    return answer